### PR TITLE
refactor(core-errors): Initial split ExtError into fallible and infallible versions

### PR DESCRIPTION
**Pull request by ark0f**
Initial `ExtError` cleanup of unreachable variants in userspace

### CODE REVIEW

File: core-backend/common/src/funcs.rs [hunk 1]
```diff
@@ -48,8 +49,9 @@ pub struct FuncsHandler<Ext: Externalities + 'static, Runtime> {
 impl<Ext, R> FuncsHandler<Ext, R>
 where
     Ext: BackendExternalities + 'static,
-    Ext::Error: BackendExternalitiesError,
-    Ext::AllocError: BackendAllocExternalitiesError<ExtError = Ext::Error>,
+    Ext::InfallibleError: BackendExternalitiesError,
```
Comment (breathx) [inline core-backend/common/src/funcs.rs:52]: I don't like `Infallible*` naming here. It's more `UnrecoverableError` that means that couldn't be caught in userspace

File: core-backend/common/src/funcs.rs [hunk 2]
```diff
@@ -297,7 +309,12 @@ where
     pub fn reply(ctx: &mut R, payload_ptr: u32, len: u32, value_ptr: u32) -> Result<(), R::Error> {
         let read_payload = ctx.register_read(payload_ptr, len);
         let value = Self::register_and_read_value(ctx, value_ptr)?;
-        let payload = ctx.read(read_payload)?.try_into()?;
+        let payload = ctx
+            .read(read_payload)?
+            .try_into()
+            .map_err(|PayloadSizeError| {
+                TrapExplanation::FallibleExt(MessageError::MaxMessageSizeExceed.into())
+            })?;
```
Comment (breathx) [inline core-backend/common/src/funcs.rs:317]: Can you please introduce this repeated stuff in ctx fn maybe?

File: core-backend/common/src/funcs.rs [hunk 3]
```diff
@@ -432,9 +454,15 @@ where
...
-
-        let s = String::from_utf8(data.into_vec())?;
+        let data: RuntimeBuffer = ctx
+            .read(read_data)?
+            .try_into()
+            .map_err(|RuntimeBufferSizeError| InfallibleMemoryError::RuntimeAllocOutOfBounds.into())
+            .map_err(TrapExplanation::InfallibleExt)?;
+
+        let s = String::from_utf8(data.into_vec())
+            .map_err(|_err| InfallibleExecutionError::InvalidDebugString.into())
```
Comment (breathx) [inline core-backend/common/src/funcs.rs:464]: double error mapping seems strange

File: core-backend/common/src/mock.rs [hunk 4]
```diff
@@ -88,7 +88,8 @@ impl CountersOwner for MockExt {
 }
 
 impl Externalities for MockExt {
```
Comment (breathx) [inline core-backend/common/src/mock.rs:90]: why are they all return Infallible?

File: core-processor/src/common.rs [hunk 5]
```diff
@@ -483,7 +483,9 @@ impl ActorExecutionErrorReplyReason {
                 TrapExplanation::GasLimitExceeded => SimpleExecutionError::RanOutOfGas,
                 TrapExplanation::ForbiddenFunction => SimpleExecutionError::BackendError,
                 TrapExplanation::ProgramAllocOutOfBounds => SimpleExecutionError::MemoryOverflow,
-                TrapExplanation::Ext(_err) => SimpleExecutionError::BackendError,
+                TrapExplanation::InfallibleExt(_) | TrapExplanation::FallibleExt(_) => {
```
Comment (breathx) [inline core-processor/src/common.rs:486]: Seems that we never can ran into FallibleExt here? :/
So trap explanation shouldn't contain this variant

**Review by breathx — CHANGES_REQUESTED** [review]

File: core-backend/common/src/funcs.rs (hunk 3 above)
Comment (ark0f) [reply core-backend/common/src/funcs.rs]: Or it will be very terrible formatted code

File: core-backend/common/src/mock.rs (hunk 4 above)
Comment (ark0f) [reply core-backend/common/src/mock.rs:90]: It's just mock

File: core-processor/src/common.rs (hunk 5 above)
Comment (ark0f) [reply core-processor/src/common.rs]: We can run either by `map_err` in `funcs.rs` or by `From` impl

File: core-backend/common/src/funcs.rs (hunk 2 above)
Comment (ark0f) [reply core-backend/common/src/funcs.rs:317]: Seems very ad-hoc-ish to me

File: core-processor/src/executor.rs [hunk 6]
```diff
@@ -259,7 +259,7 @@ pub fn execute_wasm<E>(
 where
     E: Environment,
     E::Ext: ProcessorExternalities + BackendExternalities + 'static,
-    <E::Ext as Externalities>::Error: BackendExternalitiesError,
+    <E::Ext as Externalities>::UnrecoverableError: BackendExternalitiesError,
```
Comment (SabaunT) [inline core-processor/src/executor.rs:262]: any reason for not defining same bound for `<E::Ext as Externalities>::FallibleError`?

**Review by SabaunT — APPROVED** [review]

File: core-processor/src/executor.rs (hunk 6 above)
Comment (ark0f) [reply core-processor/src/executor.rs:262]: It seems these bounds are not actually required

**Review by breathx — APPROVED** [review]

**Pull request merged**